includes/utils/class-amp-image-dimension-extractor.php

```diff
@@ -2,20 +2,44 @@
 
 class AMP_Image_Dimension_Extractor {
 	static $callbacks_registered = false;
+    const STATUS_FAILED_LAST_ATTEMPT = 'failed';
+    const STATUS_IMAGE_EXTRACTION_FAILED= 'failed';
 
-	static public function extract( $url ) {
+    static public function extract( $urls ) {
```

Comment:
To be sure, when you say "[m]aybe we can can run each image/dimension through the old filter after the fact?" you're suggesting that after we process the images in batch via my code changes, we iterate over $urls and call apply_filters( 'amp_extract_image_dimensions', false, $url ) on each one? I imagine that would require that publishers who wanted to leverage the original filter would therefore have to suppress/remove the new batch one, like how the WP.com "optimizations" currently suppress the existing callbacks on the original filter. Otherwise, if the batch filter were allowed to run first like it would by default, when the original filter kicked in for each url afterward there would be nothing for any custom callback added to it to do, since the image dimensions would have already been fetched (or marked failed) by the batch filter.

For this reason, it seems any publisher who has added custom callbacks to the original filter (and would like to see them continue to offer some utility) would have to modify code in any event to suppress the new batch filter from running first in order to ensure their custom callbacks weren't rendered inert/useless (by virtue of the batch filter already determining all image dimensions).

Where a code change would be required in any event (by a publisher who added custom callbacks to the original filter and wanted to continue to use them) by either:

A) Suppressing the new batch filter to ensure image dimensions haven't already been fetched so any callbacks added to the original filter actually have something to do; OR

B) Modifying their custom callback(s) to work with the batch filter instead of the original filter

wouldn't it be a win from a maintenance/complexity standpoint to require the latter instead of the former? Assuming a publisher was rational and only added a custom callback to the original filter that extracted dimensions from an image at a URL via some different mechanism (like any callback added to this filter is presumably supposed to), it seems it would be a non breaking change in any event since the new batch filter would simply do the image extraction work unless/until code is changed to suppress it.

Aside from the issue of the batch filter negating the usefulness of calling the original filter for each url unless/until the batch filter is explicitly suppressed/removed, for the majority of cases where publishers haven't added custom callbacks to the original filter that they're interested in maintaining, and simply leave the plugin as is, wouldn't calling the original filter for each url after the batch filter ran result in needlessly iterating over the urls in the great majority of installations/implementations?

To recap, where a code change would seemingly be required by publishers who wanted to retain a custom callback's functionality either way, it seems ideal from a maintenance/complexity standpoint not to carry the original filter forward.

If you agree, I can update wpcom-helper.php wpcom_amp_extract_image_dimensions_add_custom_callbacks() to use the batch filter as well.

If you would prefer to see the original filter invoked for each url subsequent to the invocation of the batch filter despite the above considerations, I am happy to do that as well.

Please let me know your thoughts.

Thanks,
Mike